### ISSUE TITLE
Adversarial training on models outside CLASSIFIER_LOSS_GRADIENTS_TYPE 

### ISSUE THREAD

**Issue by matthew-64** [Trusted-AI/adversarial-robustness-toolbox#991]
Hey,

**Question:**
Is there any way to adversarially train classifiers that are not in part of the `CLASSIFIER_LOSS_GRADIENTS_TYPE` but still exist in `CLASSIFIER_TYPE`?

**Exact issue:**
I am looking to adversarially train a ScikitlearnRandomForestClassifier which is part of `CLASSIFIER_TYPE` but not `CLASSIFIER_LOSS_GRADIENTS_TYPE`. When looking at the [AdversarialTrainer class](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/c600d098657111830221149f55cc6ae32e1187b1/art/defences/trainer/adversarial_trainer.py), the classifier needs to be of type `CLASSIFIER_LOSS_GRADIENTS_TYPE` as seen below:

```
class AdversarialTrainer(Trainer):

...

def __init__(
        self,
        classifier: "CLASSIFIER_LOSS_GRADIENTS_TYPE",

...
```
**Observations:**
I see evasion attacks like the [SquareAttack](https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/c600d098657111830221149f55cc6ae32e1187b1/art/attacks/evasion/square_attack.py) that will accept any classifier `CLASSIFIER_TYPE` but no adversarial trainer that will accept `CLASSIFIER_TYPE`. All the trainers will only accept `CLASSIFIER_LOSS_GRADIENTS_TYPE`. 

**Quick reference:**
```
CLASSIFIER_LOSS_GRADIENTS_TYPE = Union[
        ClassifierLossGradients,
        EnsembleClassifier,
        GPyGaussianProcessClassifier,
        KerasClassifier,
        MXClassifier,
        PyTorchClassifier,
        ScikitlearnLogisticRegression,
        ScikitlearnSVC,
        TensorFlowClassifier,
        TensorFlowV2Classifier,
    ]

    CLASSIFIER_CLASS_LOSS_GRADIENTS_TYPE = Union[
        ClassifierClassLossGradients,
        EnsembleClassifier,
        GPyGaussianProcessClassifier,
        KerasClassifier,
        MXClassifier,
        PyTorchClassifier,
        ScikitlearnLogisticRegression,
        ScikitlearnSVC,
        TensorFlowClassifier,
        TensorFlowV2Classifier,
    ]

    CLASSIFIER_NEURALNETWORK_TYPE = Union[
        ClassifierNeuralNetwork,
        DetectorClassifier,
        EnsembleClassifier,
        KerasClassifier,
        MXClassifier,
        PyTorchClassifier,
        TensorFlowClassifier,
        TensorFlowV2Classifier,
    ]

    CLASSIFIER_DECISION_TREE_TYPE = Union[
        ClassifierDecisionTree,
        LightGBMClassifier,
        ScikitlearnDecisionTreeClassifier,
        ScikitlearnDecisionTreeRegressor,
        ScikitlearnExtraTreesClassifier,
        ScikitlearnGradientBoostingClassifier,
        ScikitlearnRandomForestClassifier,
        XGBoostClassifier,
    ]

    CLASSIFIER_TYPE = Union[
        Classifier,
        BlackBoxClassifier,
        CatBoostARTClassifier,
        DetectorClassifier,
        EnsembleClassifier,
        GPyGaussianProcessClassifier,
        KerasClassifier,
        LightGBMClassifier,
        MXClassifier,
        PyTorchClassifier,
        ScikitlearnClassifier,
        ScikitlearnDecisionTreeClassifier,
        ScikitlearnDecisionTreeRegressor,
        ScikitlearnExtraTreeClassifier,
        ScikitlearnAdaBoostClassifier,
        ScikitlearnBaggingClassifier,
        ScikitlearnExtraTreesClassifier,
        ScikitlearnGradientBoostingClassifier,
        ScikitlearnRandomForestClassifier,
        ScikitlearnLogisticRegression,
        ScikitlearnSVC,
        TensorFlowClassifier,
        TensorFlowV2Classifier,
        XGBoostClassifier,
        CLASSIFIER_NEURALNETWORK_TYPE,
    ]

    OBJECT_DETECTOR_TYPE = Union[
        ObjectDetector, PyTorchFasterRCNN, TensorFlowFasterRCNN,
    ]

    SPEECH_RECOGNIZER_TYPE = Union[
        PyTorchDeepSpeech, TensorFlowLingvoASR,
    ]
```

**Comment by beat-buesser** [comment]
Hi @matthew-64

I don't think this is possible with the current trainer tools. All adversarial training tools in ART, at the moment, require loss gradients for their adversarial training algorithm and therefore require `CLASSIFIER_LOSS_GRADIENTS_TYPE` from the models they should train. `ScikitlearnRandomForestClassifier` does not provide loss gradients and is therefore not compatible with existing trainers, but it would be interesting to think of or look for an adversarial training algorithm for tree based models or models without loss gradients and add them to ART.

**Comment by matthew-64** [comment]
Hey, thanks for the swift clarification. Im really impressed by this toolbox overall.

**Comment by llamadeus** [comment]
Hey, I came across this issue during my search for an evasion attack on random forest. While Papernot's decision tree attack performed quite well, I wanted to use it for my `sklearn.ensemble.RandomForestClassifier`.

The following code snippet runs the decision tree attack for each of the random forest's decision tree. The individual decision trees are checked sequentially, and if the attack found an adversarial example which the entire ensemble misclassifies, further trees are not checked. Obviously, it takes some time to compute adversarials this way, as the attack has to be performed for many decision trees. However, I found this attack was still faster than the zeroth-order optimization attack, which was the only attack I could find for random forests.

```python
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import f1_score

from art.estimators.classification import SklearnClassifier
from art.attacks.evasion import DecisionTreeAttack

import concurrent.futures
from tqdm import tqdm
import sys

# Try to compute an adversarial from the given record
# Returns the original record, if no adversarial could be found which the whole ensemble misclassifies
def try_generate_rf_adversarial(i, model, attacks, X_one, y_true):
    for attack in attacks:
        X_adv = attack.generate(X_one.to_numpy())
        y_pred = model.predict(X_adv)

        if y_pred != y_true:
            # Use generated adversarial if model misclassified it
            return i, X_adv

    return i, X_one

# Papernot attack run for each decision tree
def rf_papernot_attack(model, X, y):
    X_adv = X.copy()
    
    classifiers = [SklearnClassifier(model=m) for m in model.estimators_]
    attacks = [DecisionTreeAttack(classifier=classifier, verbose=False) for classifier in classifiers]
    
    with concurrent.futures.ProcessPoolExecutor() as executor:
        with tqdm(total=len(X), file=sys.stdout) as progress:
            # Create a list to store the future objects
            futures = []

            # Submit the function for execution in parallel for each iteration
            for i in range(len(X)):
                record = X.iloc[[i]]
                y_actual = y.iloc[i]

                future = executor.submit(try_generate_rf_adversarial, i, model, attacks, record, y_actual)
                future.add_done_callback(lambda p: progress.update())
                futures.append(future)

            # Iterate over the completed futures to get the results
            for future in concurrent.futures.as_completed(futures):
                i, result = future.result()
                X_adv.iloc[i] = result
    
    return X_adv

# Assuming, you have X and y are pandas DataFrames and that they already exist
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

# Train the classifier
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train.to_numpy(), y_train.to_numpy())

# Predict classes of test data and evaluate f1-score
y_pred = rf.predict(X_test.to_numpy())
f1 = f1_score(y_test, y_pred)
print(f"Score on test data: {f1}")

# Compute adversarials for test data
X_adv = rf_papernot_attack(rf, X_test, y_test)

# Predict classes of adversarials and evaluate f1-score
y_pred = clf.predict(X_test.to_numpy())
f1 = f1_score(y_test, y_pred)
print(f"Score on adversarial data: {f1}")
```
